docs/rfc/active/5818-serial-updates.md

```diff
@@ -0,0 +1,313 @@
+
+## Detailed design
+
+> This is the bulk of the RFC.
+
+> Explain the design in enough detail for somebody familiar with the framework to understand, and for somebody familiar
+> with the implementation to implement. This should get into specifics and corner-cases, and include examples of how the
+> feature is used. Any new terminology should be defined here.
+
+In short, my recommendations for implementing this change is proposal 2 or 3 in the list below. But first, a 
+consideration on versions...
+
+#### Aspect Versions
+
+There are many ways to achieve this, but most of them involve DataHub aspects having some kind of consistent 
+versioning as currently the LATEST version of an aspect is version zero and is a mutable placeholder for a "real" 
+fixed version.
+
+We don't necessarily have to overhaul the versioning approach here, as all we need is a way of uniquely identifying 
+the aspect instance that is to be changed. We don't need to rely on numerical version ordering for this, but 
+something will need to be added as the general aspect level to allow for this new functionality.
+
+Whatever is done, the most important thing is that changes can be isolated at the metadata storage level, whether 
+through locking, transactions or Cassandra LWTs. More on this later.
+
+### 1: Change Processor
+
+Check the allowed set of transitions between current and proposed aspect state. This would be achieved with 
+business-specific logic injected into the DataHub codebase via plugins. This feels like teh wrong approach.
+
+Good
+* Fast as all code is executed in GMS
+
+Bad
+* No guarantee of state between the read and write
+* No guarantee that the plugin code will run
+* Specific business logic getting tied up with GMS code
+
+[![](https://mermaid.ink/img/pako:eNqFks9qAjEQxl9lyFV9gT1Yii3SgyB43cuQfLqB3USTWUXEd-_sH7vYCs0lYeY333yT5GZsdDCFyTi1CBYfng-JmzKQLrYSE61qjyBDZDgvlsvZerMr6CsckOU9H2FHQMOa3UDYsfBO61HQGk_MU3aheK_Ff1Rmq4q1wTZFi5xj0n55i9R4Ebi30WItdHzEhlDv87nyp8kVeYJCFFDyh0oo7umX58EwNXylis8g2ys6qpAwSbwcN3cbBVzojJR9DBP_evSE3Nbj6KgzyPnMdR0v_4y06KtDHCuDm26vyw6P9ZA3c9PoRbF3-ty3jiyNVGhQmkKPDnvuTJgy3BVtj2oTn86rUVPsWV3NDbcSd9dgTSGpxQMav8xI3b8BFXTCxw)](https://mermaid.live/edit#pako:eNqFks9qAjEQxl9lyFV9gT1Yii3SgyB43cuQfLqB3USTWUXEd-_sH7vYCs0lYeY333yT5GZsdDCFyTi1CBYfng-JmzKQLrYSE61qjyBDZDgvlsvZerMr6CsckOU9H2FHQMOa3UDYsfBO61HQGk_MU3aheK_Ff1Rmq4q1wTZFi5xj0n55i9R4Ebi30WItdHzEhlDv87nyp8kVeYJCFFDyh0oo7umX58EwNXylis8g2ys6qpAwSbwcN3cbBVzojJR9DBP_evSE3Nbj6KgzyPnMdR0v_4y06KtDHCuDm26vyw6P9ZA3c9PoRbF3-ty3jiyNVGhQmkKPDnvuTJgy3BVtj2oTn86rUVPsWV3NDbcSd9dgTSGpxQMav8xI3b8BFXTCxw)
+
+### 2: Read and Conditional Update
+
+Read the current aspect and pass it back to the client. The client can propose an update on the basis that what they 
+just read is still valid.
+
+Good
+* Guarantee of state between the read and write
+* No need for business-specific logic in GMS code
+
+Bad
+* Slow, especially under high contention
+* Client will need to handle retries
```

Comment:
If a conflict occurs, what should the client do? Just abort or log a warning or something, presumably?

Reply:
The client will have to manage the retry operation or just fail depending on business requirements. GMS won't have enough context to manage this without a suitable patching language that @RyanHolstien is working on.